src/hotspot/share/oops/constantPool.cpp

```diff
@@ -286,22 +294,25 @@ objArrayOop ConstantPool::prepare_resolved_references_for_archiving() {
 
   objArrayOop rr = resolved_references();
   if (rr != nullptr) {
+    ConstantPool* orig_pool = ArchiveBuilder::current()->get_source_addr(this);
     Array<u2>* ref_map = reference_map();
     int ref_map_len = ref_map == nullptr ? 0 : ref_map->length();
     int rr_len = rr->length();
     for (int i = 0; i < rr_len; i++) {
       oop obj = rr->obj_at(i);
-      rr->obj_at_put(i, nullptr);
+      HeapShared::add_scratch_resolved_reference(orig_pool, i, nullptr);
```

Comment:
I think the logic can be more obvious of the the management of the contents of the `scratch_rr` array is moved here. Something like:

```
if (rr != nullptr) {
  scratch_rr = HeapShared::scratch_resolved_references(orig_pool);
  
  .... scratch_rr->obj_at_put(i, nullptr);
  ....
  return scratch_rr;
 } 
```

That way, you won't have two functions with similar sounding names `add_scratch_resolved_reference` and `add_scratch_resolved_references`. HeapShared is now solely responsible for keeping track of the scratch arrays, but ConstantPool will be responsible for the contents of the scratch arrays.

Also, it's not necessary to put the `scratch_rr` in an `OopHandle` as we won't safepoint in this loop.